### ISSUE TITLE
Error squiggles on template template arguments

### ISSUE THREAD

**Issue by mattiasm83** [microsoft/vscode-cpptools#1014]
I fave found an issue where defaulted template type arguments in combination with template template arguments causes error squiggles that says that no instance of function template matches the argument list, even though the code is perfectly fine.

An example:

```
#include <vector>

template< template <typename...> class Container,
          typename A>
void f(const Container<A>& container)
{
}

int main(int argc, char *argv[])
{
    std::vector<int> a;
    f(a);  // <----- Problem occurs here
    return 0;
}

```
The call to f, gives me this:

> file: '/home/mattias/temp/c.cpp'
> severity: 'Error'
> message: 'no instance of function template "f" matches the argument list -- argument types are: (std::vector<int, std::allocator<int>>)'
> at: '13,5'
> source: ''

The problem seems to be the second type parameter to std::vector (the allocator), because if I change the function f to this the squiggles goes away:

```
template< template <typename...> class Container,
          typename A, typename B>
void f(const Container<A,B>& container)
{
}
```
If not using template template argument the problem goes away:

```
template<typename T>
void f(const std::vector<T>& container)
{}
```

**Comment by sean-mcmanus** [comment]
The code doesn't compile with Microsoft cl compiler, but it compiles with clang. It complains "could not deduce template argument". I don't know which compiler is correct. Can you use `template< template<typename, typename...> class Container, typename A, typename... args>
void f(const Container<A, args...>& container)` instead? That seems to be more what you're intending, because typename... could resolve to 0 template arguments, and Container<> is not possible because then the template parameter typename A could not be resolved. Another alternative would be `template< template<typename...> class Container, typename... A>
void f(const Container<A...>& container) if you really do want no-arg containers to be valid. If you believe this is a compiler bug, your best chance of getting this fixed is by reporting the bug at the Visual Studio 2017 site, because the bug repros with that product too, so the root issue is there and we could get the fix from them: https://developercommunity.visualstudio.com/ . I could also report the bug myself if you're not able to install VS 2017 (the feedback site seems to require it).

**Comment by mattiasm83** [comment]
Thanks for the reply!

I was running VSCode under linux, so I have only tried to compile with gcc and clang, and both seems to work fine. The suggestions you recommended solves the problems. I was using Vim before trying out VSCode, and I did not get the error in Vim. However, Vim probably used clang for the diagnostic messages so that might explain it.

I don't know if it is a compiler bug, only that it is allowed in clang and gcc.

**Comment by sean-mcmanus** [comment]
Thanks for reporting this. It turns out to be a bug in our compilers (for a spec discussion, see http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2016/p0522r0.html ). cl.exe has a fix in an internal build, and our IntelliSense engine is scheduled for the fix too (6 months is my best rough guess for when we could have it).

**Comment by sean-mcmanus** [comment]
FYI, our IntelliSense compiler is being updated in February, but it still doesn't have the fix for this (it might be in the next compiler update we get, but I don't know yet). One potential workaround to get rid of the error is to surround your code in
`#ifndef __INTELLISENSE__
// code that generates an error squiggle
#endif
`

**Comment by sean-mcmanus** [comment]
This appears to be fixed in our pending 0.17.7-Insiders.
